Tello3

Jakub-Bielawski opened the pull request:
I've added exemplary main function to evaluate detection and classification on web cam.

Jakub-Bielawski committed:
TELLO3 - init commit for hand gesture classification

	Added trained model, recognize fist(fist,fist_moved) and palm(palm,palm_moved)
	Add trainning script

Jakub-Bielawski committed:
TELLO3 - implement first hand classification CNN
Add:
	hand_gesture_recognition dir:
		hand_classifier
	src/test_main.py:
		for evaluation models on WEBCAM

Changed:
	src/inference/hand_detection.py
		Add model_dirs to __init__ method
		returns another image ( one to dipslay, one to feed hand_clasifier)
	src/inference/yolo.py
		eddited error handling in YOLO init

Jakub-Bielawski committed:
TELLO3 - Add trained model for hand classification
	src/models/trained_palm_fist.h5

Jakub-Bielawski committed:
TELLO3:
	update .gitignore
	dealete models dir and trained_palm_fist.h5

arekmula commented on src/hand_gesture_recognition/hand_classifier.py:
```diff
@@ -0,0 +1,37 @@
+            raise e
+
+        self.input_shape = self.model.layers[0].input_shape
+        print("Hi I'm classifier")
```
Change this to something like: "Classifier loaded successfully"

Jakub-Bielawski replied:
Yeah Yeah, it was for testing too

arekmula commented on src/hand_gesture_recognition/hand_classifier.py:
```diff
@@ -0,0 +1,37 @@
+        self.input_shape = self.model.layers[0].input_shape
+        print("Hi I'm classifier")
+
+    def predict(self, preprocessed_hand: np.array) -> np.array:
```
Change `preprocessed_hand` to `input`

arekmula commented on src/hand_gesture_recognition/hand_classifier.py:
```diff
@@ -0,0 +1,37 @@
+        image_scaled = image_grayscale / 255.
+        image_resized = cv2.resize(image_scaled, (int(self.input_shape[1]), int(self.input_shape[2])))
+        image_final = image_resized[np.newaxis, :, :, np.newaxis]
+        self.PREPROCESSED = True
```
If you run function `preprocess_input` once, then the self.PREPROCESSED will always be true, because it will be part of the class instance. You should rather do something like this:
- In function `predict` add another parameter: `should_preprocess_input`. If it will be set to True, then in the predict function run self.preprocess_input(). If it's set to False, we assume that this function's user provides with already preprocessed input.

Jakub-Bielawski replied:
Yeah, I missed it .

arekmula commented on src/hand_gesture_recognition/hand_classifier.py:
```diff
@@ -0,0 +1,37 @@
+        print("Hi I'm classifier")
+
+    def predict(self, preprocessed_hand: np.array) -> np.array:
+        if not self.PREPROCESSED:
```
```
if should_preprocess_input:
    input = self.preprocess_input()
```

arekmula commented on src/hand_gesture_recognition/notebook/Model_training.ipynb:
```diff
@@ -0,0 +1,252 @@
+{
```
Move the jupyter notebook out of `src` directory. Create separate directory called `train` and put it there

arekmula commented on src/inference/hand_detection.py:
```diff
@@ -1,14 +1,16 @@
 
-    def __init__(self, image_size=416, confidence=0.2):
-        self.yolo = YOLO("models/cross-hands-tiny.cfg", "models/cross-hands-tiny.weights", ["hand"])
+    def __init__(self, image_size, confidence, model_dirs):
```
bring back default values for image_size and confidence. model_dirs->models_dir

Jakub-Bielawski replied:
For sure

arekmula commented on src/inference/hand_detection.py:
```diff
@@ -1,14 +1,16 @@
-        self.yolo = YOLO("models/cross-hands-tiny.cfg", "models/cross-hands-tiny.weights", ["hand"])
+    def __init__(self, image_size, confidence, model_dirs):
+
+        self.yolo = YOLO(model_dirs+"cross-hands-tiny.cfg", model_dirs+"cross-hands-tiny.weights",
```
what if the user forgot to add `/` at the end of the model_dirs string?

Jakub-Bielawski replied:
I don't know how to create more robust solution, `/` has to be somewhere.

arekmula commented on src/test_main.py:
```diff
@@ -0,0 +1,82 @@
+import cv2
+
+
+def get_hand_from_img(image, boxes, enlargebox_px):
```
Move this function to hand_detector

Jakub-Bielawski replied:
Sure, it was for testing

arekmula commented on src/test_main.py:
```diff
@@ -0,0 +1,82 @@
+    return hands
+
+
+def main():
```
Delete rest of the test_main.py

Jakub-Bielawski replied:
I think it's really useful to have some easy to use example for fast developing. It's not ready yet solution and for example @Krisisss  or @Skwarson96  have to spend lass time on evaluating this for testing. In final stage of development we could remove  it or leave as "easy to use example"

arekmula replied:
Ok, that's right

Jakub-Bielawski committed:
TELLO3:
	Update:
		hand_classifier.py and test_main.py following comments
		hanf_deteciont.py added error handling while loading models
		.gitignore

arekmula commented on src/hand_gesture_recognition/hand_classifier.py:
```diff
@@ -0,0 +1,37 @@
+        image_scaled = image_grayscale / 255.
+        image_resized = cv2.resize(image_scaled, (int(self.input_shape[1]), int(self.input_shape[2])))
+        image_final = image_resized[np.newaxis, :, :, np.newaxis]
+        self.PREPROCESSED = True
```
Variable is not used, get rid of it

arekmula commented on src/inference/hand_detection.py:
```diff
@@ -1,18 +1,24 @@
 
-    def __init__(self, image_size=416, confidence=0.2):
-        self.yolo = YOLO("models/cross-hands-tiny.cfg", "models/cross-hands-tiny.weights", ["hand"])
+    def __init__(self, image_size=416, confidence=0.2, model_dirs="models/"):
```
models_dir because it's one directory with multiple models, not multiple directories.

Jakub-Bielawski committed:
TELLO3 - dealate unused variable in HandClasifier, and fix typos

Skwarson96 commented on src/inference/hand_detection.py:
```diff
@@ -26,7 +32,7 @@ def predict(self, img, should_draw_results):
         h - boz height
```
typo in box

arekmula reviewed (commented):
I forgot to mention: add new requirements to requirements.txt

Jakub-Bielawski committed:
TELLO3:
-Update requirements.txt
Added tensorflow==2.4.1

Jakub-Bielawski commented:
Ok, I'll wait for  @Skwarson96 and I'll merge
